[Python] data subscription pages batch [Replace C# samples] (#1865)
ravendb/docs: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown

| **object_type** (Optional) | `Type[_T]` | Defines the object type (class) for the items that will be included in the received `SubscriptionBatch` object. |
| **database** (Optional) | `str` | Name of the database to look for the data subscription. If `None`, the default database configured in DocumentStore will be used. |
| **subscription_name** (Optional) | `str` | The subscription's name. Used when the worker is generated without creating a `SubscriptionCreationOptions` instance, relying on the default values. |

| Return value | |
| ------------- | ----- |
| `SubscriptionWorker` | A created data subscription worker. When returned, the worker is Idle. It will start working only when its `run` function is called. |

{PANEL/}

{PANEL:`SubscriptionWorkerOptions`}

{NOTE The only mandatory parameter for the ceation of `SubscriptionWorkerOptions` is the subscription **name**. /}

| Member | Type | Description |
|--------|:-----|-------------| 
| **subscription_name** | `str` | Returns the subscription name passed to the constructor. This name will be used by the server side to identify the subscription. |

> | **time_to_wait_before_connection_retry** | `DateTime` | Time to wait before reconnecting, in the case of non-aborting failure during the subscription processing.<br>Default: 5 seconds. |

Reviewer: ```suggestion
| **time_to_wait_before_connection_retry** | `datetime` | Time to wait before reconnecting, in the case of non-aborting failure during the subscription processing.<br>Default: 5 seconds. |

```

Author: done

| **max_docs_per_batch** | `int` | Max number of documents the server will try to send in a batch. If the server doesn't find as many documents as specified, it will not wait and send those it did find.<br>Default: 4096. |
| **close_when_no_docs_left** | `bool` | If `True`, an "ad-hoc" operation processes all possible documents, until the server finds no new documents to send. When this happens, the task returned by the `run` function will fail and throw a `SubscriptionClosedException` exception.<br>Default: `False` |
| **send_buffer_size_in_bytes** | `int` | Size, in bytes, of the TCP socket buffer used for _sending_ data.<br>Default: 32,768 (32 KiB) |
| **receive_buffer_size_in_bytes** | `int` | Size, in bytes, of the TCP socket buffer used for _receiving_ data.<br>Default: 32,768 (32 KiB) |

{PANEL/}

{PANEL:Running subscription worker}

After [generating](../../../client-api/data-subscriptions/consumption/api-overview#subscription-worker-generation) 
a subscription worker, the worker doesn't process any documents yet.  
Processing starts when the `run` function is called.  

> The `run` function receives the client-side code as a delegate that will process the retrieved batches:

Reviewer: ```suggestion
The `run` function receives the client-side code as a function that will process the retrieved batches:

```

Author: done

{CODE:python subscriptionWorkerRunning@ClientApi\DataSubscriptions\DataSubscriptions.py /}

| Parameters | | |
| ------------- | ------------- | ----- |
| **process_documents** (Optional) | `[Callable[[SubscriptionBatch[_T]], Any]]` | Delegate to sync batch processing |

{PANEL/}

{PANEL:`SubscriptionBatch[_T]`}

| Member | Type | Description |
|--------|:-----|-------------| 
| **items** | `SubscriptionBatch[_T].item` array | Batch items list |
| **number_of_items_in_batch** | `int` | Number of items in the batch |

{NOTE:Subscription Worker Connectivity}
As long as there is no exception, the worker will continue addressing the same 
server that the first batch was received from.  
If the worker fails to reach that node, it will try to 
[failover](../../../client-api/configuration/load-balance/overview) to another node 
from the session's topology list.  
The node that the worker succeeds connecting to, will inform the worker which 
node is currently responsible for data subscriptions.  
{NOTE/}

> | `SubscriptionBatch[_T].item` Member | Type | Description |

Reviewer: ```suggestion
| `SubscriptionBatch[_T].Item` Member | Type | Description |

```

Author: done

|--------|:-----|-------------| 
| **result** | `T` | Current batch item |
| **exception_message** | `str` | Message of the exception thrown during current document processing in the server side |
| **id** | `str` | Current batch item's underlying document ID |
| **change_vector** | `str` | Current batch item's underlying document change vector of the current document |
| **raw_result** |  | Current batch item before serialization to `T` |
| **raw_metadata** |  | Current batch item's underlying document metadata |
| **metadata** |  | Current batch item's underlying metadata values |

{WARNING: }
Usage of `raw_result`, `raw_metadata`, and `metadata` values outside of the document processing delegate 
is not supported.
{WARNING/}

{PANEL/}

{PANEL:`SubscriptionWorker[_T]`}

---

### Methods:

| Method | Return Type | Description |
|--------|:-----|-------------| 

> | `dispose` | `void` | Aborts subscription worker operation ungracefully by waiting for the task returned by the `run` function to finish running. |

Reviewer: ```suggestion
| `close` | `void` | Aborts subscription worker operation ungracefully by waiting for the task returned by the `run` function to finish running. |

```

Author: done

> | `dispose(bool waitForSubscriptionTask)` | `void` | Aborts the subscription worker, but allows deciding whether to wait for the `run` function task or not. |

Reviewer: ```suggestion

```

> | `run` | `Task` | Starts the subscription worker work of processing batches, receiving the batch processing delegates (see [above](../../../client-api/data-subscriptions/consumption/api-overview#running-subscription-worker)). |

Reviewer: ```suggestion
| `run` | `Future[None]` | Starts the subscription worker work of processing batches, receiving the batch processing delegates (see [above](../../../client-api/data-subscriptions/consumption/api-overview#running-subscription-worker)). |

```

Author: done

---

### Events:

| Event | Type\Return type | Description |
|--------|:-----|-------------| 

> | **after\_acknowledgment** | `AfterAcknowledgmentAction` (event) | Event invoked after each time the server acknowledges batch processing progress. |

Reviewer: ```suggestion
| **after\_acknowledgment** | `Callable[[SubscriptionBatch[_T]], None]` | Event invoked after each time the server acknowledges batch processing progress. |

```

> | `AfterAcknowledgmentAction` Parameters | | |

Reviewer: maybe
```suggestion
| `Callable[[SubscriptionBatch[_T]], None]` Parameters | | |

```

but better 
```suggestion
| `after_acknowledgment` Parameters | | |

```

Author: done (the second one :o)

| ------------- | ------------- | ----- |
| **batch** | `SubscriptionBatch[_T]` | The batch process which was acknowledged |

| Return value | |
| ------------- | ----- |

> | `Task` | The worker waits for the task to finish the event processing |

Reviewer: ```suggestion
| `Future[None]` | The worker waits for the task to finish the event processing |

```

Author: done
